{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# D-SPIN scRNA-seq preprocessing\n",
    "Updated: 6/13/25.\n",
    "\n",
    "This notebook assumes the existence of top-level `data/` and `output` directories. Please see the attached `environment.yml` file in order to reproduce the environment used for this analysis. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import matplotlib\n",
    "\n",
    "import numpy as np\n",
    "import anndata as ad\n",
    "import scanpy as sc\n",
    "import pandas as pd\n",
    "\n",
    "%matplotlib inline\n",
    "sc.set_figure_params(\n",
    "    dpi=300\n",
    ")\n",
    "matplotlib.rcParams['pdf.fonttype'] = 42\n",
    "matplotlib.use('cairo')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "DATA_DIR = \"../data/gbm43_perturb_seq\"\n",
    "OUTPUT_DIR = \"../output/gbm43/dspin\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 18938 × 36601\n",
       "    obs: 'orig.ident', 'nCount_RNA', 'nFeature_RNA', 'percent.mt', 'nCount_SCT', 'nFeature_SCT', 'sorted', 'cond', 'SCT_snn_res.0.4', 'seurat_clusters', 'sgRNA_UMI', 'sgRNA_logUMI', 'sgRNA_binary', 'sgRNA', 'sgRNACond'\n",
       "    var: 'features'\n",
       "    obsm: 'X_umap'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Right now, we will load an h5ad object that contains\n",
    "# raw RNA counts. Check out what this looks like.\n",
    "gbm43_data = ad.read_h5ad(os.path.join(DATA_DIR, \"gbm43.h5ad\"))\n",
    "gbm43_data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Reprocess the data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pathlib import Path\n",
    "SAVE_PATH = Path(\"../output/gbm43\") / \"qc\"\n",
    "sc._settings.ScanpyConfig.figdir = SAVE_PATH"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Unfortunately, SCTransform doesn't work for running oNMF because there are\n",
    "# negative values produced. We will have to take the raw data, normalize it, and\n",
    "# log transform it.\n",
    "gbm43_data.var_names = gbm43_data.var_names.str.replace(\"^GRCh38-\", \"\", regex=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Do QC\n",
    "gbm43_data.var[\"mt\"] = gbm43_data.var_names.str.startswith(\"MT-\")\n",
    "gbm43_data.var[\"ribo\"] = gbm43_data.var_names.str.startswith(\"RPS\", \"RPL\")\n",
    "sc.pp.calculate_qc_metrics(\n",
    "    gbm43_data, qc_vars=[\"mt\", \"ribo\"], inplace=True, log1p=True\n",
    ")\n",
    "sc.pl.violin(\n",
    "    gbm43_data,\n",
    "    [\"n_genes_by_counts\", \"total_counts\", \"pct_counts_mt\"],\n",
    "    jitter=0.4,\n",
    "    multi_panel=True,\n",
    "    save=\"_mito_and_total_counts.pdf\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.scatter(gbm43_data, \"total_counts\", \"n_genes_by_counts\", color=\"pct_counts_mt\", save=\"_n_genes_by_counts_pct_counts_mt.pdf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 18938 × 23511\n",
       "    obs: 'orig.ident', 'nCount_RNA', 'nFeature_RNA', 'percent.mt', 'nCount_SCT', 'nFeature_SCT', 'sorted', 'cond', 'SCT_snn_res.0.4', 'seurat_clusters', 'sgRNA_UMI', 'sgRNA_logUMI', 'sgRNA_binary', 'sgRNA', 'sgRNACond', 'n_genes_by_counts', 'log1p_n_genes_by_counts', 'total_counts', 'log1p_total_counts', 'pct_counts_in_top_50_genes', 'pct_counts_in_top_100_genes', 'pct_counts_in_top_200_genes', 'pct_counts_in_top_500_genes', 'total_counts_mt', 'log1p_total_counts_mt', 'pct_counts_mt', 'total_counts_ribo', 'log1p_total_counts_ribo', 'pct_counts_ribo', 'n_genes'\n",
       "    var: 'features', 'mt', 'ribo', 'n_cells_by_counts', 'mean_counts', 'log1p_mean_counts', 'pct_dropout_by_counts', 'total_counts', 'log1p_total_counts', 'n_cells'\n",
       "    obsm: 'X_umap'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Filter permissively on mitochondrial characteristics\n",
    "sc.pp.filter_cells(gbm43_data, min_genes=100)\n",
    "sc.pp.filter_genes(gbm43_data, min_cells=3)\n",
    "\n",
    "# Note that we could implement doublet detection later\n",
    "gbm43_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Normalize and log transform the data\n",
    "gbm43_data.layers[\"counts\"] = gbm43_data.X.copy()\n",
    "sc.pp.normalize_total(gbm43_data)\n",
    "sc.pp.log1p(gbm43_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get rid of the raw object so we don't have any confusion\n",
    "gbm43_data.raw = None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Replace the cond 0s and 1s with RT/noRT\n",
    "gbm43_data.obs[\"cond\"] = gbm43_data.obs[\"cond\"].replace({0: \"noRT\", 1: \"RT\"})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gbm43_data_RT = gbm43_data[gbm43_data.obs[\"cond\"] == \"RT\"].copy()\n",
    "gbm43_data_noRT = gbm43_data[gbm43_data.obs[\"cond\"] == \"noRT\"].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# As RT and noRT may have drastically different phenotypes, we want to include highly variable genes from each of them.\n",
    "# Calculate the top 2000 highly variable genes for RT and noRT and take the union as our set of highly variable genes.\n",
    "sc.pp.highly_variable_genes(gbm43_data_RT, n_top_genes=2000, batch_key=\"orig.ident\")\n",
    "sc.pp.highly_variable_genes(gbm43_data_noRT, n_top_genes=2000, batch_key=\"orig.ident\")\n",
    "\n",
    "hvg_RT = gbm43_data_RT.var[gbm43_data_RT.var['highly_variable']].index\n",
    "hvg_noRT = gbm43_data_noRT.var[gbm43_data_noRT.var['highly_variable']].index\n",
    "\n",
    "# Take the union of highly variable genes from both conditions\n",
    "hvg_union = hvg_RT.union(hvg_noRT)\n",
    "\n",
    "# Mark the highly variable genes in the original dataset\n",
    "gbm43_data.var['highly_variable'] = gbm43_data.var_names.isin(hvg_union)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Find cell cycle scores\n",
    "cell_cycle_genes = [x.strip() for x in open(\"../data/regev_lab_cell_cycle_genes.txt\")]\n",
    "s_genes = cell_cycle_genes[:43]\n",
    "g2m_genes = cell_cycle_genes[43:]\n",
    "s_genes = [x for x in s_genes if x in gbm43_data.var_names]\n",
    "g2m_genes = [x for x in g2m_genes if x in gbm43_data.var_names]\n",
    "sc.tl.score_genes_cell_cycle(gbm43_data, s_genes=s_genes, g2m_genes=g2m_genes, use_raw=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Run PCA and clustering\n",
    "sc.tl.pca(gbm43_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WARNING: saving figure to file ../output/gbm43/qc/pca_variance_ratio_pca_variance_ratio.pdf\n"
     ]
    }
   ],
   "source": [
    "# How many PCs should we consider?\n",
    "sc.pl.pca_variance_ratio(gbm43_data, n_pcs=50, log=True, save=\"_pca_variance_ratio.pdf\", show=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WARNING: saving figure to file ../output/gbm43/qc/pca_PC1_PC2_against_samples_mt_phase.pdf\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[<AxesSubplot: title={'center': 'orig.ident'}, xlabel='PC1', ylabel='PC2'>,\n",
       " <AxesSubplot: title={'center': 'orig.ident'}, xlabel='PC3', ylabel='PC4'>,\n",
       " <AxesSubplot: title={'center': 'pct_counts_mt'}, xlabel='PC1', ylabel='PC2'>,\n",
       " <AxesSubplot: title={'center': 'pct_counts_mt'}, xlabel='PC3', ylabel='PC4'>,\n",
       " <AxesSubplot: title={'center': 'phase'}, xlabel='PC1', ylabel='PC2'>,\n",
       " <AxesSubplot: title={'center': 'phase'}, xlabel='PC3', ylabel='PC4'>]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sc.pl.pca(\n",
    "    gbm43_data,\n",
    "    color=[\"orig.ident\", \"orig.ident\", \"pct_counts_mt\", \"pct_counts_mt\", \"phase\", \"phase\"],\n",
    "    dimensions=[(0, 1), (2, 3), (0, 1), (2, 3), (0, 1), (2, 3)],\n",
    "    ncols=2,\n",
    "    size=1,\n",
    "    show=False,\n",
    "    save=\"_PC1_PC2_against_samples_mt_phase.pdf\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WARNING: saving figure to file ../output/gbm43/qc/umap_UMAP_samples.pdf\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/miniconda3/envs/human_gbm_dnapk/lib/python3.11/site-packages/scanpy/plotting/_utils.py:347: UserWarning: Matplotlib is currently using cairo, which is a non-GUI backend, so cannot show the figure.\n",
      "  plt.show()\n"
     ]
    }
   ],
   "source": [
    "# Use nearest neighbors to construct a neighborhood graph of cells using the PCA\n",
    "# dimensionality reduction\n",
    "sc.pp.neighbors(gbm43_data)\n",
    "sc.tl.umap(gbm43_data)\n",
    "sc.pl.umap(gbm43_data, color = \"orig.ident\", size = 2, save=\"_UMAP_samples.pdf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WARNING: saving figure to file ../output/gbm43/qc/umap_leiden_res_1.0.pdf\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<AxesSubplot: title={'center': 'leiden_1.0'}, xlabel='UMAP1', ylabel='UMAP2'>"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Use the Leiden graph clustering method to directly cluster the neighborhood graph.\n",
    "# Note that here we use a clustering resolution of 1.\n",
    "sc.tl.leiden(gbm43_data, resolution=1, flavor=\"igraph\", n_iterations=2, key_added=\"leiden_1.0\")\n",
    "sc.pl.umap(gbm43_data, color=[\"leiden_1.0\"], legend_loc=\"on data\", show=False, save=\"_leiden_res_1.0.pdf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WARNING: saving figure to file ../output/gbm43/qc/umap_leiden_total_counts_mt_cond_phase.pdf\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[<AxesSubplot: title={'center': 'leiden_1.0'}, xlabel='UMAP1', ylabel='UMAP2'>,\n",
       " <AxesSubplot: title={'center': 'log1p_total_counts'}, xlabel='UMAP1', ylabel='UMAP2'>,\n",
       " <AxesSubplot: title={'center': 'pct_counts_mt'}, xlabel='UMAP1', ylabel='UMAP2'>,\n",
       " <AxesSubplot: title={'center': 'log1p_n_genes_by_counts'}, xlabel='UMAP1', ylabel='UMAP2'>,\n",
       " <AxesSubplot: title={'center': 'cond'}, xlabel='UMAP1', ylabel='UMAP2'>,\n",
       " <AxesSubplot: title={'center': 'phase'}, xlabel='UMAP1', ylabel='UMAP2'>]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Visualize all of the potential sources of variance that we want/don't want\n",
    "sc.pl.umap(\n",
    "    gbm43_data,\n",
    "    color=[\"leiden_1.0\", \"log1p_total_counts\", \"pct_counts_mt\", \"log1p_n_genes_by_counts\", \"cond\", \"phase\"],\n",
    "    wspace=0.5,\n",
    "    ncols=2,\n",
    "    show=False,\n",
    "    save=\"_leiden_total_counts_mt_cond_phase.pdf\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WARNING: saving figure to file ../output/gbm43/qc/umap_leiden_resolutions.pdf\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[<AxesSubplot: title={'center': 'leiden_res_0.02'}, xlabel='UMAP1', ylabel='UMAP2'>,\n",
       " <AxesSubplot: title={'center': 'leiden_res_0.50'}, xlabel='UMAP1', ylabel='UMAP2'>,\n",
       " <AxesSubplot: title={'center': 'leiden_res_2.00'}, xlabel='UMAP1', ylabel='UMAP2'>,\n",
       " <AxesSubplot: title={'center': 'leiden_1.0'}, xlabel='UMAP1', ylabel='UMAP2'>,\n",
       " <AxesSubplot: title={'center': 'leiden_res_0.70'}, xlabel='UMAP1', ylabel='UMAP2'>]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# How many different cell types are there? It seems like somewhere between 1 and 0.5 is going to be our\n",
    "# optimal leiden clustering resolution.\n",
    "for res in [0.02, 0.5, 0.7, 2.0]:\n",
    "    sc.tl.leiden(\n",
    "        gbm43_data, key_added=f\"leiden_res_{res:4.2f}\", resolution=res, flavor=\"igraph\"\n",
    "    )\n",
    "sc.pl.umap(\n",
    "    gbm43_data,\n",
    "    color=[\"leiden_res_0.02\", \"leiden_res_0.50\", \"leiden_res_2.00\", \"leiden_1.0\", \"leiden_res_0.70\"],\n",
    "    legend_loc=\"on data\",\n",
    "    show=False,\n",
    "    save=\"_leiden_resolutions.pdf\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WARNING: Groups are not reordered because the `groupby` categories and the `var_group_labels` are different.\n",
      "categories: 0, 1, 2, etc.\n",
      "var_group_labels: low, high\n"
     ]
    }
   ],
   "source": [
    "# Calculate marker genes for the cluster of interest\n",
    "sc.tl.rank_genes_groups(\n",
    "    gbm43_data,\n",
    "    groupby=\"leiden_1.0\",\n",
    "    method=\"wilcoxon\",\n",
    "    use_raw=False\n",
    ")\n",
    "dotplot = sc.pl.rank_genes_groups_dotplot(\n",
    "    gbm43_data, groupby=\"leiden_1.0\", standard_scale=\"var\", n_genes=5, use_raw=False, show=False, save=\"_top_5_markers_leiden_1.0.pdf\", return_fig=True\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Since there are maybe 10 PCs that have the highest contribution, let's take a look at what they are\n",
    "pc_genes_dict = {}\n",
    "pca_loadings = gbm43_data.varm[\"PCs\"]\n",
    "for pc in range(10):\n",
    "    pc_genes = pca_loadings[:, pc]\n",
    "    top_genes = gbm43_data.var_names[np.argsort(-np.abs(pc_genes))[:10]].tolist()\n",
    "    pc_genes_dict[pc] = top_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WARNING: saving figure to file ../output/gbm43/qc/umap_PC_dim_0_top_genes.pdf\n",
      "WARNING: saving figure to file ../output/gbm43/qc/umap_PC_dim_1_top_genes.pdf\n",
      "WARNING: saving figure to file ../output/gbm43/qc/umap_PC_dim_2_top_genes.pdf\n"
     ]
    }
   ],
   "source": [
    "# Plot the PC genes against UMAP space\n",
    "for pc in range(3):\n",
    "    genes = pc_genes_dict[pc]\n",
    "    sc.set_figure_params(figsize=[3, 3])  # Increased figure size for better spacing\n",
    "    sc.pl.umap(\n",
    "        gbm43_data,\n",
    "        color=genes,\n",
    "        legend_loc=\"on data\",\n",
    "        show=False,\n",
    "        save=f\"_PC_dim_{pc}_top_genes.pdf\"\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WARNING: saving figure to file ../output/gbm43/qc/rank_genes_groups_umap_1_group_horizontal_UMAP_axis_top_genes.pdf\n"
     ]
    }
   ],
   "source": [
    "# What drives the horizontal axis of variation in the current UMAP plot?\n",
    "umap_1 = gbm43_data.obsm[\"X_umap\"][:, 0]\n",
    "gbm43_data.obs['umap_1_group'] = pd.cut(umap_1, bins=2, labels=['low', 'high'])\n",
    "sc.tl.rank_genes_groups(gbm43_data, groupby='umap_1_group', method='t-test', use_raw=False)\n",
    "sc.pl.rank_genes_groups(gbm43_data, n_genes=10, use_raw=False, show=False, save=\"_horizontal_UMAP_axis_top_genes.pdf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save the processed adata object\n",
    "gbm43_data.write(\n",
    "    os.path.join(DATA_DIR, \"gbm43_processed.h5ad\")\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Reload\n",
    "gbm43_data = ad.read_h5ad(os.path.join(DATA_DIR, \"gbm43_processed.h5ad\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Export the obs for manuscript purposes\n",
    "umap_df = pd.DataFrame(gbm43_data.obsm['X_umap'], index=gbm43_data.obs_names, columns=['umap_x', 'umap_y'])\n",
    "obs_with_umap_df = gbm43_data.obs.join(umap_df)\n",
    "obs_with_umap_df.index.name = \"Cell barcode\"\n",
    "obs_with_umap_df.to_csv(\"../output/manuscript/FigS2A_gbm43_perturb_seq_sc_analysis_metadata_with_umap.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [],
   "source": [
    "color_df = dotplot.dot_color_df\n",
    "size_df = dotplot.dot_size_df\n",
    "color_df.to_csv(\"../output/manuscript/FigS2B_rank_genes_groups_dotplot_color_df.csv\", index=False)\n",
    "size_df.to_csv(\"../output/manuscript/FigS2B_rank_genes_groups_dotplot_size_df.csv\", index=False)\n",
    "rank_genes_groups_info = gbm43_data.uns['rank_genes_groups']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Construct a single massive dataframe with all of this information for the top 20 ranked genes for each group\n",
    "groups = rank_genes_groups_info['names'].dtype.names\n",
    "data_records = []\n",
    "stats = ['scores', 'pvals', 'pvals_adj', 'logfoldchanges']\n",
    "for group in groups:\n",
    "    genes = rank_genes_groups_info['names'][group]\n",
    "    for rank, gene in enumerate(genes[:20]):\n",
    "        record = {'group': group, 'gene': gene, 'rank': rank}\n",
    "        for stat in stats:\n",
    "            record[stat] = rank_genes_groups_info[stat][group][rank]\n",
    "        data_records.append(record) \n",
    "data_records = pd.DataFrame(data_records)\n",
    "data_records.to_csv(\"../output/manuscript/FigS2B_rank_genes_groups_top20_genes_stats_pvalues_lfcs_dataframe.csv\", index=False)           "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Cell cycle regression"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Note that for better or worse, cell cycle regression appeared to be a significant part of the vertical variation in the UMAP.\n",
    "# Thus, we will try regressing it out. Due to downstream needs, we will not scale. In order to not regress out differences between\n",
    "# proliferative and non-proliferative cells, we will not regress out S_score and G2M_score, but instead the difference between them.\n",
    "\n",
    "# In doing so, we are constructing a dataset where we have modeled some regression coefficient for *cell_cycle_diff*. This regression\n",
    "# coefficient attempts to predict gene expression based off of some baseline + B * cell_cycle_diff for each gene, solved across cells.\n",
    "# We then subtract the influence of cell_cycle_diff between S and G2M phase cells from each cell for each gene based on the model.\n",
    "gbm43_data.obs[\"cell_cycle_diff\"] = gbm43_data.obs[\"S_score\"] - gbm43_data.obs[\"G2M_score\"]\n",
    "gbm43_data_regressed = sc.pp.regress_out(gbm43_data, [\"cell_cycle_diff\"], copy=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# For downstream analysis, we need positive values only\n",
    "sc.pp.scale(gbm43_data_regressed)\n",
    "gbm43_data_regressed.X = np.clip(gbm43_data_regressed.X, 1e-5, None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Run PCA and clustering\n",
    "sc.tl.pca(gbm43_data_regressed)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# How many PCs should we consider?\n",
    "sc.pl.pca_variance_ratio(gbm43_data_regressed, n_pcs=50, log=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.pca(\n",
    "    gbm43_data_regressed,\n",
    "    color=[\"orig.ident\", \"orig.ident\", \"pct_counts_mt\", \"pct_counts_mt\", \"phase\", \"phase\"],\n",
    "    dimensions=[(0, 1), (2, 3), (0, 1), (2, 3), (0, 1), (2, 3)],\n",
    "    ncols=2,\n",
    "    size=1,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Use nearest neighbors to construct a neighborhood graph of cells using the PCA\n",
    "# dimensionality reduction\n",
    "sc.pp.neighbors(gbm43_data_regressed)\n",
    "sc.tl.umap(gbm43_data_regressed)\n",
    "sc.pl.umap(gbm43_data_regressed, color = \"orig.ident\", size = 2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Use the Leiden graph clustering method to directly cluster the neighborhood graph.\n",
    "# Note that here we use a clustering resolution of 1.\n",
    "sc.tl.leiden(gbm43_data_regressed, resolution=1, flavor=\"igraph\", n_iterations=2, key_added=\"leiden_1.0\")\n",
    "sc.pl.umap(gbm43_data_regressed, color=[\"leiden_1.0\", \"phase\"], legend_loc=\"on data\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Since there are maybe 10 PCs that have the highest contribution, let's take a look at what they are\n",
    "pc_genes_dict = {}\n",
    "pca_loadings = gbm43_data_regressed.varm[\"PCs\"]\n",
    "for pc in range(10):\n",
    "    pc_genes = pca_loadings[:, pc]\n",
    "    top_genes = gbm43_data_regressed.var_names[np.argsort(-np.abs(pc_genes))[:10]].tolist()\n",
    "    pc_genes_dict[pc] = top_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot the PC genes against UMAP space\n",
    "for pc in range(3):\n",
    "    genes = pc_genes_dict[pc]\n",
    "    sc.set_figure_params(figsize=[3, 3])  # Increased figure size for better spacing\n",
    "    sc.pl.umap(\n",
    "        gbm43_data_regressed,\n",
    "        color=genes,\n",
    "        legend_loc=\"on data\",\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# What drives the horizontal axis of variation in the current UMAP plot?\n",
    "umap_1 = gbm43_data_regressed.obsm[\"X_umap\"][:, 0]\n",
    "gbm43_data_regressed.obs['umap_1_group'] = pd.cut(umap_1, bins=2, labels=['low', 'high'])\n",
    "sc.tl.rank_genes_groups(gbm43_data_regressed, groupby='umap_1_group', method='t-test', use_raw=False)\n",
    "sc.pl.rank_genes_groups(gbm43_data_regressed, n_genes=10, use_raw=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gbm43_data.write(\n",
    "    os.path.join(DATA_DIR, \"gbm43_processed_cell_cycle_regressed.h5ad\")\n",
    ")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "human_gbm_dnapk",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
